packages/data/src/types.ts

```diff
@@ -1,40 +1,66 @@
-export type WPDataFunctionOrGeneratorArray = Array< Function | Generator >;
-export type WPDataFunctionArray = Array< Function >;
+type MapOf< T > = { [ name: string ]: T };
 
-export interface WPDataAttachedStore {
-	getSelectors: () => WPDataFunctionArray;
-	getActions: () => WPDataFunctionArray;
+export type ActionCreator = Function | Generator;
+export type Resolver = Function | Generator;
+export type Selector = Function;
+
+export type AnyConfig = StoreConfig< any, any >;
+
+export interface StoreInstance< Config extends AnyConfig > {
+	getSelectors: () => SelectorsOf< Config >;
+	getActions: () => ActionCreatorsOf< Config >;
 	subscribe: ( listener: () => void ) => () => void;
 }
 
-export interface WPDataStore {
+export interface StoreDefinition< Config extends AnyConfig > {
 	/**
 	 * Store Name
 	 */
 	name: string;
 
 	/**
-	 * Store configuration object.
+	 * Creates a store instance
 	 */
-	instantiate: ( registry: WPDataRegistry ) => WPDataAttachedStore;
+	instantiate: ( registry: DataRegistry ) => StoreInstance< Config >;
 }
 
-export interface WPDataReduxStoreConfig {
+export interface StoreConfig<
```

Comment:
I think this should have Redux in it to clarify that it's a special case for ReduxStores and not a generic interface.

Reply:
At several points I had that in there but I took it out each time because it seemed to make things more confusing than they need to be. I can add it back, `ReduxStoreConfig`, but I hope to have a better idea in the future to replace it. At one point I made `ReduxStoreConfig extend StoreConfig` but wasn't happy with that either.

What made me withdraw from `Redux` was that the way we use these configs is pretty much the same whether it's a Redux store or not and I was thinking that if they are the same then we could keep it as one concept and pass that around.

Reply:
renamed in 0d8e9b4830